File: backend/core/src/common/access/user-scoped-crud.controller.ts
import { Body, Delete, Get, Param, Patch, Post, Req } from '@nestjs/common';
import { UserScopedCrudService } from './user-scoped-crud.service';

export class UserScopedCrudController<T extends { id: string; userId?: string }> {
  constructor(protected readonly service: UserScopedCrudService<T>) {}

  @Get()
  findAll(@Req() req: any) {
    return this.service.findAllForUser(req.user?.userId);
  }

  @Get(':id')
  findOne(@Req() req: any, @Param('id') id: string) {
    return this.service.findOneForUser(req.user?.userId, id);
  }

  @Post()
  create(@Req() req: any, @Body() dto: Partial<T>) {
    return this.service.createForUser(req.user?.userId, dto);
  }

  @Patch(':id')
  update(@Req() req: any, @Param('id') id: string, @Body() dto: Partial<T>) {
    return this.service.updateForUser(req.user?.userId, id, dto);
  }

  @Delete(':id')
  remove(@Req() req: any, @Param('id') id: string) {
    return this.service.removeForUser(req.user?.userId, id);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/common/UserScopedCrudController.java
package store.pjcloud.cowrite.core.common;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.web.bind.annotation.*;
import store.pjcloud.cowrite.core.security.SecurityUtils;

public abstract class UserScopedCrudController<T extends UserScoped> {
    protected final UserScopedCrudService<T> service;

    protected UserScopedCrudController(UserScopedCrudService<T> service) {
        this.service = service;
    }

    @GetMapping
    public List<T> findAll() {
        UUID userId = SecurityUtils.requireUserId();
        return service.findAllForUser(userId);
    }

    @GetMapping("/{id}")
    public T findOne(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        return service.findOneForUser(userId, id);
    }

    @PostMapping
    public T create(@RequestBody Map<String, Object> body) {
        UUID userId = SecurityUtils.requireUserId();
        return service.createForUser(userId, body);
    }

    @PatchMapping("/{id}")
    public T update(@PathVariable("id") UUID id, @RequestBody Map<String, Object> body) {
        UUID userId = SecurityUtils.requireUserId();
        return service.updateForUser(userId, id, body);
    }

    @DeleteMapping("/{id}")
    public Map<String, Object> remove(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        service.removeForUser(userId, id);
        return Map.of("success", true);
    }
}
